### ISSUE TITLE
[focusgroup] Why not include memory of last focused item in group?

### ISSUE THREAD

**Issue by travisleithead** [openui/open-ui#537]
Transferred from https://github.com/MicrosoftEdge/MSEdgeExplainers/issues/579

Opened by @aleventhal

Focusgroup is really neat. I'm super happy to see it being worked on, and the number of scenarios handled is amazing.

Regarding this part, I admit I'm very disappointed that focusgroup doesn't handle this for us:

> The focusgroup doesn't "remember" the last element that was focused when focus leaves the group and returns later. Instead, the "leaving" and "returning" logic (for keyboard scenarios) depends on the preexisting tabindex sequential focus navigation only (those with tabindex=0 for example). The [CSS Toggles](https://tabatkins.github.io/css-toggle/) proposal could be used in this same scenario to track the "toggle state" among this group of toggle-able elements, allowing focus changes to be decoupled from "selection/toggle" state.

Is this not being handled because of the reliance on tabindex, which already hard codes whether something is tabbable or just focusable?

Maybe all we need is another property, like focus-group-memory: on|off|value. The default would be on. When used, all focusable descendants would get a computed tabindex of 0 or -1, overriding the attribute value.

User impact: the impact would be the highest in containers with many controls.  Imagine tabbing into a grid and being at column 50, row 50. When you tab to the toolbar and come back, it would sure be helpful to be in the right place.

CC @travisleithead @benbeaudry 

@jcsteh wdyt?

----

@benbeaudry wrote:

> @travisleithead This would require a few changes in our implementation, but I do agree that the users would benefit from it. Was there any specific reason why we decided not to keep track of the last focused element when leaving a focusgroup in the first place?

---

@jcsteh responded:
 > I agree this would often be a massive efficiency benefit for users. I think it's also what users have come to expect in most cases.

**Comment by chrisdholt** [comment]
@travisleithead If I recall (and this convo was awhile ago), I think the issue here was that not every instance where someone may want to use `focusgroup` *should* automatically return someone to the previously focused instance. If someone wanted focusgroup behavior for a menu with menu-items I would not expect focusgroup to return me to the last item I invoked, I would expect it to drop me to the first focusable menu-item. 

I agree this would be a benefit to users but IMO it would likely need to be opt-in and not prescribed for all because it isn't applicable for all instances.

**Comment by aleventhal** [comment]
@jcsteh @scottaohara @benbeaudry

**Comment by aleventhal** [comment]
First, tabindex > 0 is evil, and it should be discouraged.
IMO, the only useful tabindex values are 0 (tabbable) and -1 (focusable).  When focusgroup is used everything that is focusable in any way should just be normalized to be focusable. IMO memory should be the default. The memory can clear when the focusgroup element is hidden, which takes care of the dropdown case. I'm not even 100% convinced that "non-memory" is that useful of an option to include in the first round.

**Comment by aleventhal** [comment]
@travisleithead thanks, is there a demo of that use case I can peek at?

**Comment by travisleithead** [comment]
> @travisleithead thanks, is there a demo of that use case I can peek at?

not a demo, but I think I show the use case in more detail in [this recorded OpenUI meeting](https://us02web.zoom.us/rec/play/M9CQws3fPYB6vzKjvjJi6uVgYbzHKZ2L5pWpZ-A_VNUa8SbPiqsZFoGX4TmbgClfIo2RHZdeHxoi3gyN.BJWX3lSoah5ddYDW?continueMode=true&_x_zm_rtaid=vgoo4d78TTmnS6SlqvSKQw.1631896465490.797a543ba5ef0034680f19392aa6a088&_x_zm_rhtaid=706) - around time index 04:44. Two permanent tab stops and a "roving" tab stop.

**Comment by benbeaudry** [comment]
Travis summed up pretty well the three approaches we have: reset, roving or override. 

Reset is trivial and doesn't solve this issue. It's also our current approach.

## memory-roving
I think that the memory-roving approach is interesting and is our best shot at dealing with _fixed_ tab stops within a focusgroup and keeping track of the last visited focusgroup-item (by setting its tabindex value to 0). However, I see a couple of potential issues with it:
1. All the focusgroup items above (when moving forward) a sibling focusgroup item that has a tab stop won't be "memorized" as last visited when navigating away by pressing the tab key, since that _fixed_ tab stop focusgroup-item will become the last visited focusgroup item. I think this could fall under the bad authoring category, though.
2. Having moving tab stops in-between fixed tab stops might become confusing to the end user. 

## momery-override
This approach is one that would just "make sense" for the end user, so I like a lot. Making sure that there is always only one tab stop per focusgroup and leaving it be managed by the platform is a simple and great solution.

However, I think this would render the "natively tabbable" elements (buttons, links, etc.) not tabbable anymore and would potentially be very confusing for the end user. 

Example:
```
<div focusgroup=horizontal>
    <a href=#>Link 1</a>
    <a href=#>Link 2</a>
    <a href=#>Link 3</a>
    <a href=#>Link 4</a>
</div>
```

The user might expect to be able to tab to each link, but because these links are now within a focusgroup, they aren't tabbable. This would feel weird in my opinion.

Note that for this to be a problem, the links and buttons would need to be the focusgroup items directly - not descendants of focusgroup items.

For example, this wouldn't be an issue:
```
<ul focusgroup=vertical>
    <li tabindex=-1><a href=#>Link 1</a></li>
    <li tabindex=-1><a href=#>Link 2</a></li>
    <li tabindex=-1><a href=#>Link 3</a></li>
    <li tabindex=-1><a href=#>Link 4</a></li>
</ul>
```

I think the memory-override approach is the most user-friendly, but the least developer-friendly because it breaks the tabbable behavior developers have come to expect - links, buttons, and any other element with tabindex>=0 won't be tabbable within a focusgroup. For this reason alone, I don't think this should be the default behavior - I'd like the developer to really opt-in that mode, knowing full well that it will break your usual tabindex values. This is just my opinion and should be taken with a grain of salt.

## TLDR
* memory-roving is developer friendly as it doesn't break anything and stores the last visited focusgroup item for you automagically.
* memory-roving might be confusing to users, but I might be wrong.
* memory-override is so simple it the users probably won't notice that its being used.
* memory-override can be "dangerous" if the developer uses it without reading the docs (but that never happens, right?) since it would break the tabbable behavior

## Extra question
* If a focusgroup doesn't have any tabbable focusgroup-item, should the focusgroup automatically make the first/last element tabbable (depending on the tab navigation direction)? Someone mentioned that in a comment above and I really liked the idea... should this be a default behavior regardless of "memory" mode, or would this lead to more issues?

**Comment by travisleithead** [comment]
Looking again at adding a memory (and making it the default behavior):

I do like the simplicity of the "memory-override" approach outlined above where the platform hijacks all the focusable elements, however, I think I found a problem use case that would need to be handled somehow. The scenario is: 

```html
<div focusgroup>
   <a href="somewhere">hyperlink</a>
   <label for=text>What's your name?</label>
   <input id=text type=text>
   <button>do something</button>
</div>
```

1. You Tab into the focusgroup (first time), and land on the hyperlink. focusgroup's memory is updated to the hyperlink. 
2. Then you use arrow keys to move to the input element. (Note: it would be weird for users if 'Tab' stopped working at this point.) Again, memory is updated to the input element.
3. You enter your name. Arrow keys don't work to escape the input element because it currently captures those keystrokes. (We could change input behavior to not trap the arrow keys when in a focusgroup to fix this.) You 'Tab' away. (Note, it would be weird for users if the focus didn't go to the button.)
4. Focus jumps out of the focusgroup (because the entire thing is treated like a single Tab stop).
5. Later, the user returns to the focusgroup by Shift-Tab. Focus is moved to the last-remembered item (the input element). Unless we make changes to the input element's behavior (and potentially other input trapping-element's behavior, though content-editable/EditContext might be hard to change and have compat issues), focus is now stuck in the input and the user can't get back to the hyperlink (or the button).
6. The currently focused

**Comment by travisleithead** [comment]
I'd like to propose a 4th approach, not mentioned above, that is similar to the current approach of not having a memory (and respecting all tabindex values, but which does add-in a "memory" and the memory is only used when re-entering the scope of the focusgroup (it's less complicated than "memory-roving"), and I hope it could make a good default behavior. If we need to add a way to opt-out, we could do that too.

## "entrance-memory"

In this approach, tabindex=0 and tabindex=-1 values are completely ignored **but only when focus first moves into a focusgroup**. After that, it's business-as-usual for the Tab key with respect to the document's tab order given wherever focus redirected to. So, in the [prior example](https://github.com/openui/open-ui/issues/537#issuecomment-1934738189), focus would still get trapped in the input type=text element in step 3, but step 4 would jump to the button element (next in tab order). If the user then leaves the focusgroup and returns (step 5) they would go to the last element they had focused, but could then use Tab it move around again (if they got stuck).

This approach allows webdevs to easily create roving tabindex behavior where there is only ever one tab stop, by setting everything tab-able to tabindex=-1 within a focusgroup, for example:

```html
<div focusgroup>
   <a href="somewhere" tabindex=-1>hyperlink</a>
   <label for=text>What's your name?</label>
   <input id=text type=text tabindex=-1>
   <button tabindex=-1>do something</button>
</div>
```
which, of course, re-introduces the issue outlined [above](https://github.com/openui/open-ui/issues/537#issuecomment-1934738189) where the user can get stuck in the input, but at least in that case, they introduced the problem themselves, and can fix it by making some other element in the focusgroup tab-able.

And if we consider a case that doesn't involve an element like `<input>` that captures keystrokes, then the combination of setting everything focusable to tabindex=-1 and adding a focusgroup acts just like the "memory-override" case.

Regarding @benbeaudry's [extra question](https://github.com/openui/open-ui/issues/537#issuecomment-1139104823):
> If a focusgroup doesn't have any tabbable focusgroup-item, should the focusgroup automatically make the first/last element tabbable (depending on the tab navigation direction)? Someone mentioned that in a comment above and I really liked the idea... should this be a default behavior regardless of "memory" mode, or would this lead to more issues?

I think this is nicely solved with the "entrance-memory" proposal. The first entrance to the focusgroup disregards what the tabindex values are and selects the focusable element that the focusgroup's memory is tracking, or if there is nothing in the memory, I think it should select the first focusable item in DOM order in the focusgroup (including if the first entrance is from Shift-Tab, reverse direction entrance).

**Comment by travisleithead** [comment]
Note: @gfellerph proposed an alternative approach that sticks with "memory-override", but addresses the problem of getting trapped in the input element by allowing the platform to temporarily make the next/previous focusable elements around the input element have tabindex=0 behavior to allow for a Tab-escape, and then reverting back to the normal behavior after focus moves. This would have to be pretty smart in case the next/previous elements were also focus-input traps too.

**Comment by travisleithead** [comment]
Need to also figure out under what conditions the memory of the focusgroup will be forgotten. @aleventhal [mentions](https://github.com/openui/open-ui/issues/537#issuecomment-1139021993) that one case is when the focusgroup element is hidden.
Perhaps another should be when any changes to the DOM are made to the subtree of the focusgroup element?
In any case, if the element that is being remembered gets removed from the DOM for some reason, that's at least one case to forget it. :)

**Comment by gfellerph** [comment]
It could also get disabled, inert or excluded from the focusgroup by CSS or attribute, basically if it becomes not focusable (I wish there was a browser API like `node.focusable`). 

@travisleithead, is the difference between "memory-override" and "entrance-memory" that `tabindex` settings by the author are respected the first time a focusgroup candidate gets focus but then thrown away for "memory-override" mode while "memory-override" will not initially respect the authors tabindex?

Not initialised (no element received focus yet), the author made the first button not focusable:
```html
<div focusgroup>
  <button tabindex="-1">First</button> <!-- set by the author, is skipped when user tabs into the focusgroup -->
  <input type="text">
  <input type="text">
  <input type="text">
  <button>Last</button>
</div>
```

User tabs into the focusgroup, first input receives focus as the button is being skipped:
```html
<div focusgroup>
  <button tabindex="0">First</button> <!-- gets activated for backwards tab exit -->
  <input type="text" tabindex="0"> <!-- arrow keys are now handled by the input field, roving focus is trapped -->
  <input type="text" tabindex="0"> <!-- gets activated for forwards tab exit -->
  <input type="text" tabindex="-1">
  <button tabindex="-1">Last</button>
</div>
```

In this state, it could also be possible to use `selectionStart` to determine if the cursor is at the beginning or the end of the input field and handle arrow keypresses from there on to move back to the focusgroup. This would get a little harder with `<audio>`, `<video>` and friends if they have visible controls.

After pressing tab:
```html
<div focusgroup>
  <button tabindex="-1">First</button> 
  <input type="text" tabindex="0"> <!-- gets activated for backwards tab exit -->
  <input type="text" tabindex="0"> <!-- has focus -->
  <input type="text" tabindex="0"> <!-- gets activated for forwards tab exit -->
  <button tabindex="-1">Last</button>
</div>
```

The whole focusgroup looses focus because the user clicked somewhere else:
```html
<div focusgroup>
  <button tabindex="-1">First</button> 
  <input type="text" tabindex="-1"> <!-- resets when the sibling input lost focus -->
  <input type="text" tabindex="0"> <!-- memorised -->
  <input type="text" tabindex="-1"> <!-- resets when the sibling input lost focus -->
  <button tabindex="-1">Last</button>
</div>
```

All this tabindex setting is of course just necessary if there is no other way to take an element out of focus order while keeping it interactive otherwise.

**Comment by travisleithead** [comment]
@gfellerph thanks for illustrating your idea with examples.

> @travisleithead, is the difference between "memory-override" and "entrance-memory" that tabindex settings by the author are respected the first time a focusgroup candidate gets focus but then thrown away for "memory-override" mode while "memory-override" will not initially respect the authors tabindex?

Did you mean 'but then thrown away for ***"entrance memory"*** mode while "memory-override" will not initially respect the authors tabindex?'? I'm not sure that quite captures it. Let me clarify using your example setup. (And remember that focusgroup will not actually change tabindex values in the DOM--I assume your examples are illustrative of what is happening internally).

From the not-initialized state (no memory yet), a tab into the focusgroup would put the focus on the `<button>` in your example in both "entrance-memory" and "memory-override". 
* In "entrance-memory" this is because the "entrance" to the focusgroup just picks the memory element OR the first focusable thing (ignoring tabindex=-1) in DOM order when nothing is in the memory. A tab key press would then go to the first `<input>` and so on like normal.
* in "memory-override" the `<button>` is selected for the same reason, and then it works like a typical roving tabindex set. A tab key press would then jump out of the focusgroup.

Hope that helps.

**Comment by travisleithead** [comment]
Current plan:
* change the explainer to add a "memory" to the focusgroup with the least disruption/magic to existing Tab behavior as possible.  A roving-tab-index "memory" is a nearly universal feature of all existing roving tab-index patterns. The MVP of this feature would not be complete without it. The ["entrance-memory" proposal](https://github.com/openui/open-ui/issues/537#issuecomment-1934930124) adds a memory that has minimum impact on other Tab behavior allowing developers to workaround potential issues noted above with focus-trapping elements without needing to add other magic behaviors.

**Comment by travisleithead** [comment]
Note from @kbrilla copied here from #990 
 
> `Roving tabindex with focusgroup (take 2)` (from [this doc](https://docs.google.com/document/d/1cOxjjZdvFfpexiN1x-MpNJLXrmbrR0xp1sRCTbaQ29c/))
Needs ability to opt out from memory mechanism as for example common pattern is using tabindex=0 on selected option and when you tab into/tab back into you always want that selected item to be first focused not the last item where you were (for example using arrows). From what I understand this will not be possible as memory mechanism takes priority before tabindex=0

I think this makes sense and is a good use case for having an opt-out.

**Comment by travisleithead** [comment]
PR #1021 has landed and includes the changes for an "entrance-memory" as [described above](https://github.com/openui/open-ui/issues/537#issuecomment-1957736539). That PR is a major re-write, so might be worth reading the explainer again top-to-bottom. Note, it also includes definition of a memory opt-out via `no-memory` or `focus-group-memory: none`.
